casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeployUtil.scala

```diff
@@ -0,0 +1,47 @@
+package coop.rchain.casper.util.rholang
+
+import com.google.protobuf.ByteString
+
+import coop.rchain.casper.genesis.contracts.Validator
+import coop.rchain.casper.protocol.DeployData
+import coop.rchain.crypto.hash.Blake2b512Random
+import coop.rchain.crypto.signatures.Signed
+import coop.rchain.crypto.PublicKey
+
+object SystemDeployUtil {
+  import coop.rchain.models.Validator.Validator
+  // Currently we have 4 system deploys -> refund, preCharge, closeBlock, Slashing
+  // In every user deploy, the rnode would do the preCharge first, then execute the
+  // user deploy and do the refund at last.
+  //
+  // The refund and preCharge system deploy
+  // would use user deploy signature to generate the system deploy. The random seed of
+  // the refund and preCharge has to be exactly the same to make sure replay the user
+  // deploy would come out the exact same result.
+  //
+  // As for closeBlock and slashing, the rnode would execute closeBlock system deploy in every block.
+  // The closeBlock system deploy would use the sender and the seqNum of the validator.
+
+  def generateSystemDeployRandomSeed(sender: Validator, seqNum: Int): Blake2b512Random =
+    Tools.rng(sender.toByteArray() ++ sender.toByteArray())
+
+  // splitByte here to make sure random seed is not the same as precharge deploy when there
+  // is only one user deploy in a block
+  def generateCloseDeployRandomSeed(sender: Validator, seqNum: Int): Blake2b512Random =
+    generateSystemDeployRandomSeed(sender, seqNum).splitByte(1)
+
+  def generateCloseDeployRandomSeed(pk: PublicKey, seqNum: Int): Blake2b512Random = {
+    val sender = ByteString.copyFrom(pk.bytes)
+    generateSystemDeployRandomSeed(sender, seqNum).splitByte(1)
+  }
+
+  def generateSlashDeployRandomSeed(sender: Validator, seqNum: Int): Blake2b512Random =
+    generateSystemDeployRandomSeed(sender, seqNum).splitByte(2)
+
+  def generatePreChargeDeployRandomSeed(deploy: Signed[DeployData]): Blake2b512Random =
+    Tools.rng(deploy.sig.toByteArray)
+
+  def generateRefundDeployRandomSeed(deploy: Signed[DeployData]): Blake2b512Random =
+    Tools.rng(deploy.sig.toByteArray).splitByte(64)
```

Comment:
This `64` is like magic number, maybe we can use split index `0` for pre-charge and `1` for refund.